File: Binary Tree Right Side View/Binary_Tree_Right_Side_View.cpp
// 使用广度优先搜索返回每一层中最后的一个节点的值
// ms %

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include<queue>
#include<vector>
using namespace std;

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> values;
        queue<TreeNode*> treeQueue;
        treeQueue.push(root);
        
        if (!root)
            return values;

        while (!treeQueue.empty())
        {
            int size = treeQueue.size();
            while (--size >= 1)
            {
                root = treeQueue.front();
                treeQueue.pop();

                if (root->left) treeQueue.push(root->left);
                if (root->right) treeQueue.push(root->right);
            }

            root = treeQueue.front();
            treeQueue.pop();

            values.push_back(root->val);

            if (root->left) treeQueue.push(root->left);
            if (root->right) treeQueue.push(root->right);
        }
        return values;
    }
};

File: Binary Tree Right Side View/Binary_Tree_Right_Side_View.py
# 60ms 99.56%

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        values = []
        treeQueue = []
        treeQueue.append(root)

        if root is None:
            return values
        while len(treeQueue) > 0:
            size = len(treeQueue)
            while size >= 2:
                root = treeQueue.pop(0)
                if root.left:
                    treeQueue.append(root.left)
                if root.right:
                    treeQueue.append(root.right)
            
            root = treeQueue.pop(0)
            values.append(root.val)
            if root.left:
                treeQueue.append(root.left)
            if root.right:
                treeQueue.append(root.right)
                size -= 1
        return values